data-structures.md

```diff
@@ -107,24 +109,45 @@ type Actor struct {
 	Head    Cid
     
     // Nonce is a counter of the number of messages this actor has sent
-	Nonce   Integer
+	Nonce   Uint64
     
     // Balance is this actors current balance of filecoin
-	Balance AttoFIL
+	Balance BigInteger
 }
 ```
 
+## State Tree
 
+The state trie keeps track of all state in Filecoin. It is effectively a map of addresses to `actors` in the system. It is implemented using a HAMT.
 
+## HAMT
 
-### Serialization
+TODO: link to spec for our CHAMP HAMT
 
-Actors are currently serialized simply by CBOR marshaling them, using lower-camel-cased field names.
 
-## State Tree
 
-The state trie keeps track of all state in Filecoin. It is effectively a map of addresses to `actors` in the system. It is implemented using a HAMT.
+# Filecoin Compact Serialization
 
-## HAMT
+Datastructures in Filecoin are encoded as compactly as is reasonable. At a high level, each object is converted into an ordered array of its fields (ordered by their appearance in the struct declaration), then CBOR marshaled, and prepended with an object type tag.
+
+| object | tag  |
+|---|---|
```

Comment:
implementers will need a bunch of examples to ensure their implementation is working. uh, including us i suppose. also prolly a good idea to link to something that makes it easy to decode cbor to diagnostic output.

Reply:
I'll link this somewhere too, but [this tool](https://github.com/polydawn/refmt/blob/master/cmd/refmt/main.go) is really great for debug printing cbor, or even converting it to json

Reply:
- [ ] TODO: add examples and link diagnostics tool.